docs/langref/relay_type.rst

```diff
@@ -0,0 +1,222 @@
+`(Tensor[(), bool], Tensor[(10, 10), float32])`
+and :code:`%c` is of type `Tensor[(10, 10), float32]`.
+
+.. code-block:: python
+   let %t = (Constant(0, (), bool), Constant(1, (10, 10), float32));
+   let %c = %t.1;
+   %c
+
+See :py:class:`~tvm.relay.ty.TupleType` for its definition and documentation.
+
+Type Parameter
+~~~~~~~~~~~~~~
+
+Type parameters represent placeholder types used for polymorphism in functions.
+Type parameters are specified according to *kind*: Kinds include general types
+(such as tensor types and tuple types), shapes, and data types. Relay enforces
+that types can only appear where their kind permits them; for instance, a
+shape or data type can only appear inside a tensor type, while only general types
+may appear inside a tensor or for a function's return and argument types.
+
+Type parameters will be substituted with a concrete type at call sites.
+
+For example, `s` below is a type paramter of kind `Shape` and it will
+be substituted for `(10, 10)` at the call site below:
+
+.. code-block:: python
+
+   def @plus<s : Shape>(%t1 : Tensor[s, float32], %t2 : Tensor[s, float32]) {
+        add(%t1, %t2)
+   }
+   plus<(10, 10)>(%a, %b)
+
+
+See :py:class:`~tvm.relay.ty.TypeVar` for its definition and documentation.
+
+Type Constraint
+~~~~~~~~~~~~~~~
+
+Abstract class representing a type constraint, to be elaborated
+upon in further releases. Currently, type relations are the only
+type constraints provided; they are discussed below.
+
+See :py:class:`~tvm.relay.ty.TypeConstraint` for its definition and documentation.
+
+Function Type
+~~~~~~~~~~~~~
+
+A function type in Relay, see `tvm/relay/type.h` for more details.
+
+This is the type assigned to functions in Relay. A function type
+consists of a list of type parameters, a set of type constraints,
+a sequence of argument types, and a return type.
+
+We informally write function types as:
+`fn<type_params>(arg_types) -> ret_type where type_constraints`
+
+A type parameter in the function type may appear in the argument
+types or the return types. Additionally, each of the type constraints
+must hold at every call site of the function. The type constraints
+typically take the function's argument types and the function's return
+type as arguments, but may take a subset instead.
+
+See :py:class:`~tvm.relay.ty.FuncType` for its definition and documentation.
+
+Type Relation
+~~~~~~~~~~~~~
+
+A type relation is the most complex type system feature in Relay.
+It allows users to extend type inference with new rules.
+We use type relations to define types for operators that work with
+tensor shapes in complex ways, such as broadcasting operators or
+:code:`flatten`, allowing Relay to statically reason about the shapes
+in these cases.
+
+A type relation :code:`R` is an n-ary-input, single-output relation over
+types. Namely, :code:`R` specifies a relationship between its argument
+types and outputs `true` if the relationship holds and `false`
```

Comment:
```suggestion
outputs `true` if the relationship holds and `false`
```

Reply:
what happens if there is not enough information to determine whether or not the relation holds?

Reply:
The relation would still be `true` but incomplete types or shape vars would remain. I'll make a note about that.